dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/NormalizeLoops.java

```diff
@@ -70,4 +116,57 @@ public void run(Function function) {
             counter++;
         }
     }
+
+    private void guaranteeSingleEntry(Function function) {
+        int labelCounter = 0;
+        for (Label label : function.getEvents(Label.class)) {
+
+            final List<CondJump> backJumps = label.getJumpSet().stream()
+                    .filter(j -> j.getLocalId() > label.getLocalId())
+                    .sorted()
+                    .toList();
+            if (backJumps.isEmpty()) {
+                continue;
+            }
+
+            int counter = 0;
+
+            final Label loopBegin = label;
+            final CondJump last = backJumps.get(backJumps.size() - 1);
+
+            final Register sourceReg = function.newRegister(String.format("__jumpedTo_%s#%s_From", label, labelCounter), types.getArchType());
+            final Local initJumpedFromOutside = EventFactory.newLocal(sourceReg, expressions.makeZero(types.getArchType()));
+            function.getEntry().insertAfter(initJumpedFromOutside);
+
+            final List<Label> loopBodyLabels = loopBegin.getSuccessor().getSuccessors().stream()
+                    .takeWhile(ev -> ev != last)
+                    .filter(Label.class::isInstance).map(Label.class::cast)
+                    .toList();
+
+            for (Label l : loopBodyLabels) {
+                final List<CondJump> externalEntries = l.getJumpSet().stream()
+                        .filter(j -> j.getLocalId() < loopBegin.getLocalId())
+                        .toList();
+
+                for (CondJump fromOutside : externalEntries) {
+                    counter++;
```

Comment:
I think you are confusing `counter` (from "where" externally we are jumping) with `labelCounter` (which loop are we jumping in). Notice that `counter` is not used to the register name but as the values that are assigned to the register, i.e.
```
// labelCounter = 1, counter = 0
jumpingTo1 <- 0 // init
...
// labelCounter = 1, counter = 1
// replacing jumping from outside
jumpingTo1 <- 1 
goto 1:
...
1:
```

Reply:
I understand and that's why I said "should get the same value" (not same register or anything). But if you have two external jumps `jumpingTo1_1` and `jumpingTo1_2` that both target the same inner label, then both should generate the exact same value after the transformation.
```
jumpingTo1 = 1;
goto loopBegin // Jump number one to <innerLabel>
...
jumpingTo1 = 1; // Same value because we go to the same inner label
gotoLoopBegin // Different jump but also targets <innerLabel>

loopBegin:
if (jumpingTo1 == 1) goto innerLabel; // Only one case for two external jumps.
...
innerLabel:
...
```

So for each inner label you need a fresh value and not for each jump that goes to some inner label.